nova/src/ivc.rs

```diff
@@ -1,135 +1,327 @@
+            u_single: Some(u_single_next_primary),
+            u_range: Some(self.u_range_primary.clone()),
+            commit_t: Some(commit_t_primary),
+            f: Default::default(),
         };
 
-        let mut cs = R1cs::<C>::default();
-        augmented_circuit.generate(&mut cs);
-
-        let (u_single_next, w_single_next) = (
-            RelaxedR1csInstance::new(DenseVectors::new(cs.x())),
-            RelaxedR1csWitness::new(DenseVectors::new(cs.w()), self.r1cs.m()),
-        );
+        let zi_secondary = circuit_secondary.generate(&mut cs_secondary);
+        let (u_single_next_secondary, w_single_next_secondary) =
+            r1cs_instance_and_witness(&cs_secondary, &pp.r1cs_shape_secondary, &pp.ck_secondary);
 
-        assert_eq!(u_single_next.x.len(), 1);
-        assert_eq!(u_single_next.x[0], u_single_next_x);
-
-        self.u_single = self.u_single.update(&u_single_next, &w_single_next);
-        self.u_range = self.u_range.update(&u_range_next, &w_range_next);
+        // update values
         self.i += 1;
-        self.zi = z_next;
-
-        // ((Ui+1, Wi+1), (ui+1, wi+1))
-        let pair = (
-            (self.u_range.instance.clone(), self.u_range.witness.clone()),
-            (
-                self.u_single.instance.clone(),
-                self.u_single.witness.clone(),
-            ),
+        self.u_range_primary = u_range_next_primary;
+        self.w_range_primary = w_range_next_primary;
+        self.u_range_secondary = u_range_next_secondary;
+        self.w_range_secondary = w_range_next_secondary;
+        self.u_single_secondary = u_single_next_secondary;
+        self.w_single_secondary = w_single_next_secondary;
+        self.zi_primary = DenseVectors::new(
+            zi_primary
+                .into_iter()
+                .map(|x| x.value(&cs_primary))
+                .collect(),
+        );
+        self.zi_secondary = DenseVectors::new(
+            zi_secondary
+                .into_iter()
+                .map(|x| x.value(&cs_secondary))
+                .collect(),
         );
 
         RecursiveProof {
             i: self.i,
-            z0: self.z0.clone(),
-            zi: self.zi.clone(),
-            r1cs: self.r1cs.clone(),
-            pair,
+            z0_primary: self.z0_primary.clone(),
+            z0_secondary: self.z0_secondary.clone(),
+            zi_primary: self.zi_primary.clone(),
+            zi_secondary: self.zi_secondary.clone(),
+            instances: (
+                (
+                    self.u_single_secondary.clone(),
+                    self.w_single_secondary.clone(),
+                ),
+                (self.u_range_primary.clone(), self.w_range_primary.clone()),
+                (
+                    self.u_range_secondary.clone(),
+                    self.w_range_secondary.clone(),
+                ),
+            ),
+            marker: Default::default(),
+        }
+    }
+}
+
+pub struct PublicParams<E1, E2, FC1, FC2>
```

Comment:
Shapes for primary and secondary augmented circuits. Doesn't change the form and can be used on any iteration.
Commitment keys (passed to provers, used for r1cs verification)